## EXPLORING TORONTO Neighborhood - PART 3

Here we will use the **explore** function to get the most common venue categories in each neighborhood, and then use this feature to group the neighborhoods into clusters using the *k*-means clustering algorithm.

Before we get the data and start exploring it, let's download all the dependencies that we will need.

In [68]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

# !conda install -c conda-forge folium=0.5.0 --yes 
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [69]:
toronto_data = pd.read_csv('toronto_comb.csv',index_col=0)
toronto_data.head()

Postal Code      Borough                          Neighborhood   Latitude  \
0         M1B  Scarborough                         Rouge,Malvern  43.806686   
1         M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2         M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3         M1G  Scarborough                                Woburn  43.770992   
4         M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [70]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(toronto_data['Borough'].unique()),
        toronto_data.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


<a id='item1'></a>


#### Use geopy library to get the latitude and longitude values of Toronto

In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>can_explorer</em>, as shown below.

In [71]:
address = 'Toronto, ON'
geolocator = Nominatim(user_agent="tor_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


#### Create a map of Toronto with neighborhoods superimposed on top.

In [72]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Borough'], toronto_data['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### Define Foursquare Credentials and Version

In [73]:
CLIENT_ID = '5LJIZQISPWRP4Q5H5QGZB5BM410UVH0RCPPGYWJD3KQXMQF1' # your Foursquare ID
CLIENT_SECRET = 'EGIMPUJ3N0HAPGNZ4UOF2N3MN1TP4L3T2TK0JM0UNHIIFV1C' # your Foursquare Secret
VERSION = '20180604'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 5LJIZQISPWRP4Q5H5QGZB5BM410UVH0RCPPGYWJD3KQXMQF1
CLIENT_SECRET:EGIMPUJ3N0HAPGNZ4UOF2N3MN1TP4L3T2TK0JM0UNHIIFV1C


<a id='item2'></a>

## Explore Neighborhoods in Toronto

#### Let's create a function to explore all the venues within 500m of each neighborhood in Toronto

In [74]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called *toronto_venues*.

In [75]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude'],
                                  )

Rouge,Malvern
Highland Creek,Rouge Hill,Port Union
Guildwood,Morningside,West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park,Ionview,Kennedy Park
Clairlea,Golden Mile,Oakridge
Cliffcrest,Cliffside,Scarborough Village West
Birch Cliff,Cliffside West
Dorset Park,Scarborough Town Centre,Wexford Heights
Maryvale,Wexford
Agincourt
Clarks Corners,Sullivan,Tam O'Shanter
Agincourt North,L'Amoreaux East,Milliken,Steeles East
L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview,Henry Farm,Oriole
Bayview Village
Silver Hills,York Mills
Newtonbrook,Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park,Don Mills South
Bathurst Manor,Downsview North,Wilson Heights
Northwood Park,York University
CFB Toronto,Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens,Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West,Riverdale
The Beaches West,Indi

#### Let's check the size of the resulting dataframe

In [76]:
print(toronto_venues.shape)
toronto_venues.head()

(1338, 7)


Neighborhood  Neighborhood Latitude  \
0                         Rouge,Malvern              43.806686   
1  Highland Creek,Rouge Hill,Port Union              43.784535   
2  Highland Creek,Rouge Hill,Port Union              43.784535   
3       Guildwood,Morningside,West Hill              43.763573   
4       Guildwood,Morningside,West Hill              43.763573   

   Neighborhood Longitude                      Venue  Venue Latitude  \
0              -79.194353                    Wendy’s       43.807448   
1              -79.160497      Royal Canadian Legion       43.782533   
2              -79.160497  Affordable Toronto Movers       43.787919   
3              -79.188711             RBC Royal Bank       43.766790   
4              -79.188711          G & G Electronics       43.765309   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.163085                   Bar  
2       -79.162977         Moving Target  
3       -79.191151                  Bank  
4       -79.191537     Electronics Store

Let's check how many venues were returned for each neighborhood

In [77]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide,King,Richmond                                                 30   
Agincourt                                                               5   
Agincourt North,L'Amoreaux East,Milliken,Steele...                      3   
Albion Gardens,Beaumond Heights,Humbergate,Jame...                     11   
Alderwood,Long Branch                                                  10   
Bathurst Manor,Downsview North,Wilson Heights                          19   
Bayview Village                                                         4   
Bedford Park,Lawrence Manor East                                       23   
Berczy Park                                                            30   
Birch Cliff,Cliffside West                                              4   
Bloordale Gardens,Eringate,Markland Wood,Old Bu...                      8   
Brockton,Exhibition Place,Parkdale Village                             22   
Business Reply Mail Processing Centre 969 Eastern                      15   
CFB Toronto,Downsview East                                              3   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                     17   
Cabbagetown,St. James Town                                             30   
Caledonia-Fairbanks                                                     4   
Canada Post Gateway Processing Centre                                  12   
Cedarbrae                                                               9   
Central Bay Street                                                     30   
Chinatown,Grange Park,Kensington Market                                30   
Christie                                                               17   
Church and Wellesley                                                   30   
Clairlea,Golden Mile,Oakridge                                          10   
Clarks Corners,Sullivan,Tam O'Shanter                                  14   
Cliffcrest,Cliffside,Scarborough Village West                           2   
Cloverdale,Islington,Martin Grove,Princess Gard...                      1   
Commerce Court,Victoria Hotel                                          30   
Davisville                                                             30   
Davisville North                                                        8   
Deer Park,Forest Hill SE,Rathnelly,South Hill,S...                     16   
Del Ray,Keelesdale,Mount Dennis,Silverthorn                             5   
Design Exchange,Toronto Dominion Centre                                30   
Don Mills North                                                         5   
Dorset Park,Scarborough Town Centre,Wexford Hei...                      7   
Dovercourt Village,Dufferin                                            19   
Downsview Central                                                       3   
Downsview Northwest                                                     4   
Downsview West                                                          4   
Downsview,North Park,Upwood Park                                        3   
East Birchmount Park,Ionview,Kennedy Park                               5   
East Toronto                                                            4   
Emery,Humberlea                                                         1   
Fairview,Henry Farm,Oriole                                             30   
First Canadian Place,Underground city                                  30   
Flemingdon Park,Don Mills South                                        23   
Forest Hill North,Forest Hill West                                      4   
Glencairn                                                               4   
Guildwood,Morningside,West Hill                                         7   
Harbord,University of Toronto                                          30   
Harbourfront                                    

#### Let's find out how many unique categories can be curated from all the returned venues

In [78]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 228 uniques categories.


<a id='item3'></a>

##  Analyze Each Neighborhood

In [79]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0         0            0           0                    0                 0   
1         0            0           0                    0                 0   
2         0            0           0                    0                 0   
3         0            0           0                    0                 0   
4         0            0           0                    0                 0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                   0              0          0           0           0   
1                   0              0          0           0           0   
2                   0              0          0           0           0   
3                   0              0          0           0           0   
4                   0              0          0           0           0   

   Bakery  Bank  Bar  Baseball Field  Basketball Stadium  Beer Bar  \
0       0     0    0               0                   0         0   
1       0     0    1               0                   0         0   
2       0     0    0               0                   0         0   
3       0     1    0               0                   0         0   
4       0     0    0               0                   0         0   

   Beer Store  Belgian Restaurant  Bike Shop  Bistro  Boat or Ferry  \
0           0                   0          0       0              0   
1           0                   0          0       0              0   
2           0                   0          0       0              0   
3           0                   0          0       0              0   
4           0                   0          0       0              0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0          0         0                     0               0        0   
1          0         0                     0               0        0   
2          0         0                     0               0        0   
3          0         0                     0               0        0   
4          0         0                     0               0        0   

   Bridal Shop  Bubble Tea Shop  Burger Joint  Burrito Place  Bus Line  \
0            0                0             0              0         0   
1            0                0             0              0         0   
2            0                0             0              0         0   
3            0                0             0              0         0   
4            0                0             0              0         0   

   Bus Station  Butcher  Cafeteria  Café  Cajun / Creole Restaurant  \
0            0        0          0     0                          0   
1            0        0          0     0                          0   
2            0        0          0     0                          0   
3            0        0          0     0                          0   
4            0        0          0     0      

And let's examine the new dataframe size.

In [80]:
toronto_onehot.shape

(1338, 228)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [81]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                              Adelaide,King,Richmond     0.000000   
1                                           Agincourt     0.000000   
2   Agincourt North,L'Amoreaux East,Milliken,Steel...     0.000000   
3   Albion Gardens,Beaumond Heights,Humbergate,Jam...     0.000000   
4                               Alderwood,Long Branch     0.000000   
5       Bathurst Manor,Downsview North,Wilson Heights     0.000000   
6                                     Bayview Village     0.000000   
7                    Bedford Park,Lawrence Manor East     0.000000   
8                                         Berczy Park     0.000000   
9                          Birch Cliff,Cliffside West     0.000000   
10  Bloordale Gardens,Eringate,Markland Wood,Old B...     0.000000   
11         Brockton,Exhibition Place,Parkdale Village     0.000000   
12  Business Reply Mail Processing Centre 969 Eastern     0.000000   
13                         CFB Toronto,Downsview East     0.000000   
14  CN Tower,Bathurst Quay,Island airport,Harbourf...     0.000000   
15                         Cabbagetown,St. James Town     0.000000   
16                                Caledonia-Fairbanks     0.000000   
17              Canada Post Gateway Processing Centre     0.000000   
18                                          Cedarbrae     0.000000   
19                                 Central Bay Street     0.033333   
20            Chinatown,Grange Park,Kensington Market     0.000000   
21                                           Christie     0.000000   
22                               Church and Wellesley     0.000000   
23                      Clairlea,Golden Mile,Oakridge     0.000000   
24              Clarks Corners,Sullivan,Tam O'Shanter     0.000000   
25      Cliffcrest,Cliffside,Scarborough Village West     0.000000   
26  Cloverdale,Islington,Martin Grove,Princess Gar...     0.000000   
27                      Commerce Court,Victoria Hotel     0.000000   
28                                         Davisville     0.000000   
29                                   Davisville North     0.000000   
30  Deer Park,Forest Hill SE,Rathnelly,South Hill,...     0.000000   
31        Del Ray,Keelesdale,Mount Dennis,Silverthorn     0.000000   
32            Design Exchange,Toronto Dominion Centre     0.000000   
33                                    Don Mills North     0.000000   
34  Dorset Park,Scarborough Town Centre,Wexford He...     0.000000   
35                        Dovercourt Village,Dufferin     0.000000   
36                                  Downsview Central     0.000000   
37                                Downsview Northwest     0.000000   
38                                     Downsview West     0.000000   
39                   Downsview,North Park,Upwood Park     0.000000   
40          East Birchmount Park,Ionview,Kennedy Park     0.000000   
41                                       East Toronto     0.000000   
42                                    Emery,Humberlea     0.000000   
43                         Fairview,Henry Farm,Oriole     0.000000   
44              First Canadian Place,Underground city     0.000000   
45                    Flemingdon Park,Don Mills South     0.000000   
46                 Forest Hill North,Forest Hill West     0.000000   
47                                          Glencairn     0.000000   
48                    Guildwood,Morningside,West Hill     0.000000   
49                      Harbord,University of Toronto     0.033333   
50                                       Harbourfront     0.033333   
51    Harbourfront East,Toronto Islands,Union Station     0.000000   
52                       High Park,The Junction South     0.000000   
53               Highland Creek,Rouge Hill,Port Union     0.000000   
54                                  Hillcrest Village     0.000000   
55         Humber Bay Shores,Mimico South,New Toronto     0.000000   
56  Humber Bay,King's Mill Park,Kingsway Park Sout.

#### Let's confirm the new size

In [82]:
toronto_grouped.shape

(99, 228)

#### Let's print each neighborhood along with the top 5 most common venues

In [83]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide,King,Richmond----
                venue  freq
0                Café  0.10
1         Coffee Shop  0.10
2  Seafood Restaurant  0.07
3               Plaza  0.03
4          Smoke Shop  0.03


----Agincourt----
                       venue  freq
0                     Lounge   0.2
1             Breakfast Spot   0.2
2  Latin American Restaurant   0.2
3             Clothing Store   0.2
4               Skating Rink   0.2


----Agincourt North,L'Amoreaux East,Milliken,Steeles East----
         venue  freq
0         Park  0.33
1   Playground  0.33
2  Coffee Shop  0.33
3  Yoga Studio  0.00
4        Motel  0.00


----Albion Gardens,Beaumond Heights,Humbergate,Jamestown,Mount Olive,Silverstone,South Steeles,Thistletown----
                  venue  freq
0         Grocery Store  0.18
1              Pharmacy  0.09
2           Coffee Shop  0.09
3  Fast Food Restaurant  0.09
4   Fried Chicken Joint  0.09


----Alderwood,Long Branch----
                venue  freq
0         Pizza Place   0.2


                        venue  freq
0                        Park  0.33
1  Construction & Landscaping  0.33
2                      Bakery  0.33
3                 Yoga Studio  0.00
4         Monument / Landmark  0.00


----East Birchmount Park,Ionview,Kennedy Park----
              venue  freq
0       Coffee Shop   0.2
1     Train Station   0.2
2       Bus Station   0.2
3  Department Store   0.2
4    Discount Store   0.2


----East Toronto----
               venue  freq
0               Park  0.50
1        Coffee Shop  0.25
2  Convenience Store  0.25
3        Yoga Studio  0.00
4              Motel  0.00


----Emery,Humberlea----
                 venue  freq
0       Baseball Field   1.0
1          Yoga Studio   0.0
2  Monument / Landmark   0.0
3    Martial Arts Dojo   0.0
4       Massage Studio   0.0


----Fairview,Henry Farm,Oriole----
            venue  freq
0     Coffee Shop  0.17
1  Clothing Store  0.13
2      Restaurant  0.07
3            Bank  0.07
4          Bakery  0.03


----Firs

         venue  freq
0         Café  0.10
1    Gastropub  0.10
2  Coffee Shop  0.07
3          Gym  0.03
4      Theater  0.03


----Stn A PO Boxes 25 The Esplanade----
                venue  freq
0            Beer Bar  0.07
1  Seafood Restaurant  0.07
2        Cocktail Bar  0.07
3          Restaurant  0.07
4                Café  0.07


----Studio District----
             venue  freq
0             Café  0.13
1      Coffee Shop  0.10
2           Bakery  0.07
3      Yoga Studio  0.03
4  Coworking Space  0.03


----The Annex,North Midtown,Yorkville----
                       venue  freq
0                       Café  0.12
1             Sandwich Place  0.12
2                Coffee Shop  0.08
3          Indian Restaurant  0.04
4  Middle Eastern Restaurant  0.04


----The Beaches----
               venue  freq
0  Health Food Store  0.25
1                Pub  0.25
2              Trail  0.25
3        Yoga Studio  0.00
4              Motel  0.00


----The Beaches West,India Bazaar----
          

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [84]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [85]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                             Adelaide,King,Richmond           Coffee Shop   
1                                          Agincourt                Lounge   
2  Agincourt North,L'Amoreaux East,Milliken,Steel...           Coffee Shop   
3  Albion Gardens,Beaumond Heights,Humbergate,Jam...         Grocery Store   
4                              Alderwood,Long Branch           Pizza Place   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                       Café    Seafood Restaurant        General Travel   
1  Latin American Restaurant        Breakfast Spot        Clothing Store   
2                 Playground                  Park         Women's Store   
3        Japanese Restaurant           Coffee Shop   Fried Chicken Joint   
4                        Gym              Pharmacy          Skating Rink   

  5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
0                Lounge                        Hotel             Speakeasy   
1          Skating Rink                 Dessert Shop     Electronics Store   
2          Dance Studio  Eastern European Restaurant             Drugstore   
3              Pharmacy                  Pizza Place  Fast Food Restaurant   
4        Sandwich Place                          Pub                  Pool   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                   Smoke Shop   Japanese Restaurant             Restaurant  
1  Eastern European Restaurant             Drugstore             Donut Shop  
2                   Donut Shop               Dog Run    Distribution Center  
3               Sandwich Place            Beer Store         Discount Store  
4           Athletics & Sports           Coffee Shop             Comic Shop

<a id='item4'></a>

##  Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [86]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 1, 2, 0, 0, 0, 0, 0, 0, 0])

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [87]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!
#neighborhoods_venues_sorted.head()

Postal Code      Borough                          Neighborhood   Latitude  \
0         M1B  Scarborough                         Rouge,Malvern  43.806686   
1         M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2         M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3         M1G  Scarborough                                Woburn  43.770992   
4         M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.194353             3.0  Fast Food Restaurant    Falafel Restaurant   
1 -79.160497             1.0                   Bar         Moving Target   
2 -79.188711             1.0        Breakfast Spot        Medical Center   
3 -79.216917             0.0           Coffee Shop     Korean Restaurant   
4 -79.239476             1.0    Athletics & Sports           Gas Station   

  3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
0           Event Space     Electronics Store  Eastern European Restaurant   
1         Women's Store      Department Store                  Event Space   
2                  Bank          Intersection           Mexican Restaurant   
3         Women's Store      Department Store                  Event Space   
4                  Bank                Bakery          Fried Chicken Joint   

  6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
0             Drugstore                   Donut Shop               Dog Run   
1     Electronics Store  Eastern European Restaurant             Drugstore   
2   Rental Car Location            Electronics Store         Women's Store   
3     Electronics Store  Eastern European Restaurant             Drugstore   
4  Caribbean Restaurant              Thai Restaurant                Lounge   

  9th Most Common Venue 10th Most Common Venue  
0   Distribution Center         Discount Store  
1            Donut Shop                Dog Run  
2          Dessert Shop              Drugstore  
3            Donut Shop                Dog Run  
4      Hakka Restaurant        Coworking Space

Finally, let's visualize the resulting clusters

In [101]:
import math
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[0],
        fill=True,
        fill_color=rainbow[2],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

<a id='item5'></a>

## Examine Clusters

Examine each cluster and determine the discriminating venue categories that distinguish each cluster.

#### Cluster 1

In [102]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels  1st Most Common Venue  \
3         Scarborough             0.0            Coffee Shop   
6         Scarborough             0.0       Department Store   
8         Scarborough             0.0    American Restaurant   
9         Scarborough             0.0        College Stadium   
13        Scarborough             0.0            Pizza Place   
15        Scarborough             0.0   Fast Food Restaurant   
18         North York             0.0            Coffee Shop   
19         North York             0.0    Japanese Restaurant   
22         North York             0.0       Ramen Restaurant   
24         North York             0.0                Butcher   
26         North York             0.0     Athletics & Sports   
27         North York             0.0                    Gym   
28         North York             0.0            Coffee Shop   
29         North York             0.0     Falafel Restaurant   
34         North York             0.0  Portuguese Restaurant   
35          East York             0.0            Pizza Place   
38          East York             0.0            Coffee Shop   
43       East Toronto             0.0                   Café   
46    Central Toronto             0.0         Clothing Store   
47    Central Toronto             0.0           Dessert Shop   
49    Central Toronto             0.0                    Pub   
51   Downtown Toronto             0.0                 Bakery   
53   Downtown Toronto             0.0            Coffee Shop   
54   Downtown Toronto             0.0                   Café   
55   Downtown Toronto             0.0                   Café   
56   Downtown Toronto             0.0           Cocktail Bar   
57   Downtown Toronto             0.0            Coffee Shop   
58   Downtown Toronto             0.0            Coffee Shop   
60   Downtown Toronto             0.0            Coffee Shop   
61   Downtown Toronto             0.0                   Café   
62         North York             0.0              Juice Bar   
65    Central Toronto             0.0         Sandwich Place   
66   Downtown Toronto             0.0                   Café   
67   Downtown Toronto             0.0                   Café   
69   Downtown Toronto             0.0           Cocktail Bar   
70   Downtown Toronto             0.0                   Café   
75   Downtown Toronto             0.0          Grocery Store   
78       West Toronto             0.0                   Café   
80               York             0.0            Coffee Shop   
83       West Toronto             0.0              Gift Shop   
84       West Toronto             0.0                   Café   
85   Downtown Toronto             0.0            Coffee Shop   
86        Mississauga             0.0                  Hotel   
88          Etobicoke             0.0            Coffee Shop   
89          Etobicoke             0.0            Pizza Place   
95          Etobicoke             0.0                   Park   
99          Etobicoke             0.0            Pizza Place   
101         Etobicoke             0.0          Grocery Store   

      2nd Most Common Venue      3rd Most Common Venue  \
3         Korean Restaurant              Women's Store   
6               Coffee Shop              Train Station   
8                     Motel              Women's Store   
9     General Entertainment               Skating Rink   
13              Coffee Shop        Fried Chicken Joint   
15       Chinese Restaurant                   Pharmacy   
18           Clothing Store                 Restaurant   
19       Chinese Restaurant                       Bank   
22           Sandwich Place                Coffee Shop   
24              Coffee Shop                   Pharmacy   
26                     Café       Gym / Fitness Center   
27         Asian Restaurant                Coffee Shop   
28                     Bank  Middle Eastern Restaurant   
29   Furniture / Home Store                Coffee Shop   
34     

#### Cluster 2

In [103]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels    1st Most Common Venue  \
1         Scarborough             1.0                      Bar   
2         Scarborough             1.0           Breakfast Spot   
4         Scarborough             1.0       Athletics & Sports   
5         Scarborough             1.0               Playground   
7         Scarborough             1.0                 Bus Line   
10        Scarborough             1.0        Indian Restaurant   
11        Scarborough             1.0               Smoke Shop   
12        Scarborough             1.0                   Lounge   
17         North York             1.0              Golf Course   
20         North York             1.0                Cafeteria   
31         North York             1.0            Grocery Store   
32         North York             1.0           Baseball Field   
33         North York             1.0            Grocery Store   
36          East York             1.0                     Park   
37       East Toronto             1.0        Health Food Store   
39          East York             1.0        Indian Restaurant   
41       East Toronto             1.0         Greek Restaurant   
42       East Toronto             1.0                     Park   
44    Central Toronto             1.0                     Park   
45    Central Toronto             1.0                      Gym   
48    Central Toronto             1.0             Tennis Court   
52   Downtown Toronto             1.0                Bookstore   
59   Downtown Toronto             1.0                    Plaza   
63    Central Toronto             1.0  Health & Beauty Service   
64    Central Toronto             1.0            Jewelry Store   
68   Downtown Toronto             1.0          Airport Service   
71         North York             1.0           Clothing Store   
72         North York             1.0                     Park   
73               York             1.0                    Field   
76       West Toronto             1.0                 Pharmacy   
77       West Toronto             1.0                      Bar   
81               York             1.0            Grocery Store   
82       West Toronto             1.0       Mexican Restaurant   
87       East Toronto             1.0       Light Rail Station   
91          Etobicoke             1.0            Deli / Bodega   
92          Etobicoke             1.0                      Gym   
97         North York             1.0           Baseball Field   
100         Etobicoke             1.0                     Park   
102         Etobicoke             1.0                Drugstore   

          2nd Most Common Venue  3rd Most Common Venue  \
1                 Moving Target          Women's Store   
2                Medical Center                   Bank   
4                   Gas Station                   Bank   
5    Construction & Landscaping          Women's Store   
7                        Bakery                   Park   
10           Light Rail Station     Chinese Restaurant   
11               Breakfast Spot                 Bakery   
12    Latin American Restaurant         Breakfast Spot   
17         Fast Food Restaurant                Dog Run   
20            Martial Arts Dojo     Falafel Restaurant   
31                         Bank                   Park   
32                   Food Truck           Home Service   
33           Athletics & Sports         Discount Store   
36             Asian Restaurant               Pharmacy   
37                        Trail                    Pub   
39                          Gym          Grocery Store   
41               Ice Cream Shop     Italian Restaurant   
42         Fast Food Restaurant                    Gym   
44                         Lake            Swim School   
45                        Hotel           Dance Studio   
48                   Playground          Women's Store   
52                  Pizza Place       Sushi Restaurant   
59                        Hotel        

#### Cluster 3

In [104]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14       Scarborough             2.0           Coffee Shop   
23        North York             2.0                  Park   
25        North York             2.0                  Park   
30        North York             2.0                  Park   
40         East York             2.0                  Park   
50  Downtown Toronto             2.0                  Park   
74              York             2.0                  Park   
79        North York             2.0                  Park   
90         Etobicoke             2.0                  Park   
98              York             2.0                  Park   

         2nd Most Common Venue 3rd Most Common Venue  \
14                  Playground                  Park   
23           Convenience Store                  Bank   
25  Construction & Landscaping     Food & Drink Shop   
30                     Airport           Snack Place   
40                 Coffee Shop     Convenience Store   
50                  Playground                 Trail   
74                        Pool         Women's Store   
79  Construction & Landscaping                Bakery   
90                       River          Dance Studio   
98          Falafel Restaurant     Electronics Store   

          4th Most Common Venue        5th Most Common Venue  \
14                Women's Store                 Dance Studio   
23             Department Store                  Event Space   
25                Deli / Bodega            Electronics Store   
30                Deli / Bodega            Electronics Store   
40             Department Store                  Event Space   
50                  Curling Ice  Eastern European Restaurant   
74         Colombian Restaurant      Comfort Food Restaurant   
79             Department Store                  Event Space   
90            Electronics Store  Eastern European Restaurant   
98  Eastern European Restaurant                    Drugstore   

          6th Most Common Venue        7th Most Common Venue  \
14  Eastern European Restaurant                    Drugstore   
23            Electronics Store  Eastern European Restaurant   
25  Eastern European Restaurant                    Drugstore   
30  Eastern European Restaurant                    Drugstore   
40            Electronics Store  Eastern European Restaurant   
50                    Drugstore                   Donut Shop   
74            Electronics Store  Eastern European Restaurant   
79            Electronics Store  Eastern European Restaurant   
90                    Drugstore                   Donut Shop   
98                   Donut Shop                      Dog Run   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
14            Donut Shop               Dog Run    Distribution Center  
23             Drugstore            Donut Shop                Dog Run  
25            Donut Shop               Dog Run    Distribution Center  
30            Donut Shop               Dog Run    Distribution Center  
40             Drugstore            Donut Shop                Dog Run  
50               Dog Run   Distribution Center         Discount Store  
74             Drugstore            Donut Shop                Dog Run  
79             Drugstore            Donut Shop                Dog Run  
90               Dog Run   Distribution Center         Discount Store  
98   Distribution Center        Discount Store                  Diner

#### Cluster 4

In [105]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0  Scarborough             3.0  Fast Food Restaurant    Falafel Restaurant   

  3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
0           Event Space     Electronics Store  Eastern European Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0             Drugstore            Donut Shop               Dog Run   

  9th Most Common Venue 10th Most Common Venue  
0   Distribution Center         Discount Store

#### Cluster 5

In [106]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
94  Etobicoke             4.0           Golf Course         Women's Store   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
94         Deli / Bodega     Electronics Store  Eastern European Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
94             Drugstore            Donut Shop               Dog Run   

   9th Most Common Venue 10th Most Common Venue  
94   Distribution Center         Discount Store

From the segmented clusters it can very well be noted that:

* Clusters 1 and 2 are the largest with 101 and 102 neighbourhoods each.
* Clusters 4 and 5 are the smallest with only 1 neighbourhood each.
* Cluster 1 has lot of cafe, restaurants and other places of leisure.
* Cluster 2 has a mix of everything like eateries, grocery, pharmacy,electronic store,gym etc..
* Cluster 3 has lot of parks, gym, sports grounds, health clubs etc.
* Cluster 4 has different types of restaurants as the most common venue
* Cluster 5 has golf course and women store as the main venues.


Thus we have analyzed and segmented the Toronto neighborhoods using clustering technique and found some interesting facts. 

Thank you!